### PR TITLE
Integrate interop with Crashlytics sdk

### PR DESCRIPTION
**Pull request by themiswang**
Integrate RC interop with Crashlytics SDK:
1. Adding Crashlytics subscriber for RC interop
2. Adding Deferred Proxy  for RC interop
3. Unit tests

### CODE REVIEW

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/RemoteConfigDeferredProxy.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  public void setupListener(UserMetadata userMetadata) {
+    this.userMetadata.set(userMetadata);
+    remoteConfigInteropDeferred.whenAvailable(
+        remoteConfigInteropProvider -> {
+          CrashlyticsRemoteConfigListener listener =
+              new CrashlyticsRemoteConfigListener(this.userMetadata.get());
+          FirebaseRemoteConfigInterop interop = remoteConfigInteropProvider.get();
+          interop.registerRolloutsStateSubscriber("crashlytics", listener);
```
Comment (samedson) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/RemoteConfigDeferredProxy.java:38]: Could we add a debug log here so we know when it's subscribed to help with debugging issues?

File: firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/CrashlyticsRemoteConfigListenerTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class CrashlyticsRemoteConfigListenerTest {
```
[the remark this answers is not in the document]
Comment (themiswang) [reply firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/CrashlyticsRemoteConfigListenerTest.java:31]: Originally the reason to also add json init is for us to reuse the class from interop. But in the end we consider each team maintain its own class. When interop pass us the`RolloutsState` object, it should already contain a list valid rolloutAssignment. I was cheat on the initialization a bit. But I can do more than one assignment case

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/CrashlyticsRemoteConfigListener.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      RolloutAssignment crashlyticsRolloutAssignment =
+          RolloutAssignment.create(
+              assignment.getRolloutId(),
+              assignment.getParameterKey(),
+              assignment.getParameterValue(),
+              assignment.getVariantId(),
+              assignment.getTemplateVersion());
+      crashlyticsRolloutAssignmentList.add(crashlyticsRolloutAssignment);
+    }
+    userMetadata.updateRolloutsState(crashlyticsRolloutAssignmentList);
```
Comment (samedson) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/CrashlyticsRemoteConfigListener.java:51]: Could we also add a debug log here? Something like "Updated Crashlytics Rollout State"

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/RemoteConfigDeferredProxy.java (hunk 1 above)
Comment (samedson) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/RemoteConfigDeferredProxy.java:38]: Could be something like, "Registering RemoteConfig Rollouts subscriber"

**Review by samedson — APPROVED** [review]
This looks great 🚀

File: firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/CrashlyticsRemoteConfigListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (samedson) [reply firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/CrashlyticsRemoteConfigListenerTest.java:31]: Ah that makes sense

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import androidx.annotation.NonNull;
+import com.google.firebase.crashlytics.internal.metadata.RolloutAssignment;
+import com.google.firebase.crashlytics.internal.metadata.UserMetadata;
+import com.google.firebase.remoteconfig.interop.rollouts.RolloutsState;
+import com.google.firebase.remoteconfig.interop.rollouts.RolloutsStateSubscriber;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+public class CrashlyticsRemoteConfigListener implements RolloutsStateSubscriber {
```
Comment (danasilver) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:26]: Super nit: might call the superclass name "...Subscriber"

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  CrashlyticsRemoteConfigListener(UserMetadata userMetadata) {
+    this.userMetadata = userMetadata;
+  }
+
+  @Override
+  public void onRolloutsStateChanged(@NonNull RolloutsState rolloutsState) {
+    Set<com.google.firebase.remoteconfig.interop.rollouts.RolloutAssignment> rolloutAssignmentList =
+        rolloutsState.getRolloutAssignments();
+
+    List<com.google.firebase.crashlytics.internal.metadata.RolloutAssignment>
```
Comment (danasilver) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:38]: I think you can just use the unqualified `RolloutAssignment` here like below

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java (hunk 4 above)
Comment (themiswang) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:26]: Ahhh the reason for Listener is for Analytics integration we call it `Listener`. Want to keep the naming consistency on our side

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java (hunk 5 above)
Comment (themiswang) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:38]: One `RolloutAssignment` is from interop and another is Crashlytics class. Unfortunately in Java you cannot import same class name from different source

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/RemoteConfigDeferredProxy.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    if (userMetadata.get() == null) {
+      Logger.getLogger().w("Didn't successfully register with UserMetadata for rollouts listener");
+      return;
+    }
+    CrashlyticsRemoteConfigListener listener =
+        new CrashlyticsRemoteConfigListener(userMetadata.get());
+    remoteConfigInteropDeferred.whenAvailable(
+        remoteConfigInteropProvider -> {
+          FirebaseRemoteConfigInterop interop = remoteConfigInteropProvider.get();
+          interop.registerRolloutsStateSubscriber("crashlytics", listener);
```
Comment (danasilver) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/RemoteConfigDeferredProxy.java:42]: `registerRolloutsStateSubscriber` should just take the listener as a parameter

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java (hunk 5 above)
Comment (danasilver) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:38]: Right right. I _think_ this is the one imported from Crashlytics (and referenced below without the full package)?

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java (hunk 4 above)
Comment (danasilver) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:26]: Ah. Maybe we want to just update the Interop naming. We already have `ConfigUpdateListener` for RC Realtime in the RC SDK...

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/RemoteConfigDeferredProxy.java (hunk 6 above)
Comment (themiswang) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/RemoteConfigDeferredProxy.java:42]: Seems like need to touch interop, shall we address it in a followup pr?

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java (hunk 4 above)
Comment (themiswang) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:26]: Seems like need to touch interop, shall we address it in a followup pr?

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java (hunk 5 above)
Comment (themiswang) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/CrashlyticsRemoteConfigListener.java:38]: Indeed!

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/RemoteConfigDeferredProxy.java (hunk 6 above)
Comment (themiswang) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/RemoteConfigDeferredProxy.java:42]: Seems I have to add the namespace:
Screenshot 2023-08-30 at 10 47 33 PM

**Review by mrober — APPROVED** [review]

**Review by danasilver — APPROVED** [review]

**Pull request merged**